### ISSUE TITLE
Investigate pause/resume problems in p2's tests

### ISSUE THREAD

**Issue by laeubi** [eclipse-equinox/p2#301]
I have tried to find out where/if pause/resume is actually used anywhere, the only reference I can found is in

`org.eclipse.equinox.internal.p2.engine.PhaseSet.pause()`

but this code is actually only called in test code.

@merks lets collect here any insights about pause/resume ...

**Comment by merks** [comment]
I looked at the call hierarchy and found uses in these tests:

![image](https://github.com/eclipse-equinox/p2/assets/208716/7782ae0b-ce6c-4489-9430-ac032e112ebc)

I think the phase stuff isn't related.  This appears to be the related framework code:

org.eclipse.equinox.internal.p2.transport.ecf.RepositoryTransport.download(URI, OutputStream, IProgressMonitor)

And these the tests that presumably hang:

org.eclipse.equinox.p2.tests.repository.FileReaderTest2

**Comment by laeubi** [comment]
> I think the phase stuff isn't related. 

If you look who calls `FileReader#pause()` you will find that a listener is waiting for `DownloadPauseResumeEvent.TYPE_PAUSE` event, this is fired from `org.eclipse.equinox.internal.p2.engine.phases.Collect.setPaused(boolean)` what is called from `org.eclipse.equinox.internal.p2.engine.PhaseSet.pause()`

**Comment by merks** [comment]
Ah yes, that's a relation indeed.  

Also, I can reproduce the hang with the debug launched test:

![image](https://github.com/eclipse-equinox/p2/assets/208716/c198ae0b-7d12-4099-8b1f-619fa88df048)

**Comment by merks** [comment]
It's annoying but I don't have source code for the new implementation.   It seems that some guards in that implementation are not calling "up" in order to fire the resume event:

![image](https://github.com/eclipse-equinox/p2/assets/208716/7a1d8649-1d17-4796-b489-c07273c0f011)

I think I need to figure out how to get the source in there too...

**Comment by laeubi** [comment]
> It's annoying but I don't have source code for the new implementation. 

I think you probably need to import the project from the ECF repository and then add the project to the source lookup from the debug configuration.

**Comment by merks** [comment]
Yes, Oomph did a good job with adding the dependencies directly from the Git clone easily...

The following change fixes the test failure:

![image](https://github.com/eclipse-equinox/p2/assets/208716/569882c6-b88b-4fb4-8f3a-bd5732df4f3b)

I.e., the new implementation considers the "transferable" to be still connected while paused (because there is an httpResponse and it's not canceled).  The old implementation was like this so does return isConnected true for this test:
```
	protected boolean isConnected() {
		return (getMethod != null);
	}
```

**Comment by merks** [comment]
Of course this "fix" is kind of fishy...

**Comment by merks** [comment]
In any case, I need to run soon and will be away from a computer until later in the afternoon....

**Comment by laeubi** [comment]
@merks I'll try to take a look at the impl but must confess its quite confusing, I think the logic somehow tries to account for cases where resume is called multiple times (same for pause) but if it is connected or not seems not a good indication as possibly the connection can take some time.

**Comment by merks** [comment]
Yes, it's indeed all a bit fishy, but neither of us probably want to spend a lot of time on it and given it's not really generally used in "our" applications, it probably doesn't much matter so as long as the tests pass it's probably good enough...

**Comment by merks** [comment]
BTW, when testing the installer, it's not finding any registered implementation and is falling back to a URL connection.  That doesn't seem right and I will need to debug further to determine what needs to change in the installer to use this new stuff properly.

**Comment by merks** [comment]
@laeubi 

I got the installer working with the new httpclientjava feature, and mostly everything just works fine.   Oomph directly uses ECF has well and the hack we did to support cookie handling doesn't work, but I'll investigate that separately.   The only serious problem I noticed evident in this stack trace:
```
org.eclipse.core.runtime.CoreException: Unable to read repository at https://download.eclipse.org/stats/oomph/feature/2.3.0/org.eclipse.oomph.setup.
	at org.eclipse.equinox.internal.p2.transport.ecf.RepositoryTransport.getLastModified(RepositoryTransport.java:283)
	at org.eclipse.oomph.p2.internal.core.CachingTransport.delegateGetLastModified(CachingTransport.java:415)
	at org.eclipse.oomph.p2.internal.core.CachingTransport.getLastModified(CachingTransport.java:364)
	at org.eclipse.equinox.internal.p2.artifact.repository.MirrorRequest.collectStats(MirrorRequest.java:279)
	at org.eclipse.equinox.internal.p2.artifact.repository.MirrorRequest.transfer(MirrorRequest.java:243)
	at org.eclipse.equinox.internal.p2.artifact.repository.MirrorRequest.perform(MirrorRequest.java:162)
	at org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository.getArtifact(SimpleArtifactRepository.java:775)
	at org.eclipse.equinox.internal.p2.artifact.repository.simple.SimpleArtifactRepository.getArtifacts(SimpleArtifactRepository.java:855)
	at org.eclipse.equinox.internal.p2.artifact.repository.CompositeArtifactRepository.getArtifacts(CompositeArtifactRepository.java:362)
	at org.eclipse.equinox.internal.p2.engine.DownloadManager.fetch(DownloadManager.java:127)
	at org.eclipse.equinox.internal.p2.engine.DownloadManager.start(DownloadManager.java:98)
	at org.eclipse.equinox.internal.p2.engine.phases.Collect.completePhase(Collect.java:124)
	at org.eclipse.equinox.internal.p2.engine.Phase.postPerform(Phase.java:254)
	at org.eclipse.equinox.internal.p2.engine.Phase.perform(Phase.java:105)
	at org.eclipse.equinox.internal.p2.engine.PhaseSet.perform(PhaseSet.java:50)
	at org.eclipse.equinox.internal.p2.engine.Engine.perform(Engine.java:95)
	at org.eclipse.equinox.internal.p2.engine.Engine.perform(Engine.java:49)
	at org.eclipse.equinox.internal.provisional.p2.director.PlanExecutionHelper.executePlan(PlanExecutionHelper.java:48)
	at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl$3.commit(ProfileTransactionImpl.java:574)
	at org.eclipse.oomph.p2.internal.core.ProfileTransactionImpl.commit(ProfileTransactionImpl.java:359)
	at org.eclipse.oomph.setup.p2.impl.P2TaskImpl.perform(P2TaskImpl.java:904)
	at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.doPerformNeededSetupTasks(SetupTaskPerformer.java:3864)
	at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performNeededSetupTasks(SetupTaskPerformer.java:3792)
	at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.performTriggeredSetupTasks(SetupTaskPerformer.java:3773)
	at org.eclipse.oomph.setup.internal.core.SetupTaskPerformer.perform(SetupTaskPerformer.java:3651)
	at org.eclipse.oomph.setup.ui.wizards.ProgressPage$9.run(ProgressPage.java:592)
	at org.eclipse.oomph.setup.ui.wizards.ProgressPage$11$1.run(ProgressPage.java:721)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.io.IOException: HTTP/1.1 header parser received no bytes
	at java.net.http/jdk.internal.net.http.common.Utils.wrapWithExtraDetail(Utils.java:348)
	at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.onReadError(Http1Response.java:675)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.checkForErrors(Http1AsyncReceiver.java:302)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:268)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230)
	at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:157)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:305)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:274)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.onReadError(Http1AsyncReceiver.java:511)
	at java.net.http/jdk.internal.net.http.Http1AsyncReceiver$Http1TubeSubscriber.onComplete(Http1AsyncReceiver.java:596)
	at java.net.http/jdk.internal.net.http.common.SSLTube$DelegateWrapper.onComplete(SSLTube.java:276)
	at java.net.http/jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.complete(SSLTube.java:440)
	at java.net.http/jdk.internal.net.http.common.SSLTube$SSLSubscriberWrapper.onComplete(SSLTube.java:541)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper.checkCompletion(SubscriberWrapper.java:472)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run1(SubscriberWrapper.java:334)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper$DownstreamPusher.run(SubscriberWrapper.java:259)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:303)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler.runOrSchedule(SequentialScheduler.java:256)
	at java.net.http/jdk.internal.net.http.common.SubscriberWrapper.outgoing(SubscriberWrapper.java:232)
	at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader.processData(SSLFlowDelegate.java:513)
	at java.net.http/jdk.internal.net.http.common.SSLFlowDelegate$Reader$ReaderDownstreamPusher.run(SSLFlowDelegate.java:268)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:205)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149)
	at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:230)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.EOFException: EOF reached while reading
	... 21 more
```

Don't be distracted by the specialized transport, the problem is that this expects a FileNotFoundException:

![image](https://github.com/eclipse-equinox/p2/assets/208716/26c375de-9e60-426f-b001-b3d6dbe2d757)

But that's not what's thrown, so there is some behavior difference that creates noise.

I'll debug a bit further to see what's different.

**Comment by merks** [comment]
In a fresh user install I noticed more problems in Oomph.  Specifically the old implementation has Map<String, String>:

```
	private Map<String, String> getResponseHeaders() {
		if (getMethod == null)
			return null;
		Header[] headers = httpResponse.getHeaders();
		Map<String, String> result = null;
		if (headers != null && headers.length > 0) {
			result = new HashMap<String, String>();
			for (Header header : headers) {
				String name = header.getName();
				String val = header.getValue();
				if (name != null && val != null)
					result.put(name, val);
			}
		}
		return Collections.unmodifiableMap(result);
	}
```
But the new implementation has Map<String, List<String>> which makes it not an exact drop-in replacement:
```
	private Map<String, List<String>> getResponseHeaders() {
		if (httpResponse == null || httpResponse.isCancelled())
			return null;
		return Collections.unmodifiableMap(getHeader().map());
	}
```
All of which is not evident nor documented by IIncomingFileTransferReceiveStartEvent

```
	public Map getResponseHeaders();
```
But I can work around that...

**Comment by merks** [comment]
This is really hard to debug.  If I drop to frame and run again, it produces a FileNotFoundException the second time.  So the behavior is timing dependent.  The ones that fail are different each time.   Given the exception is created here:

![image](https://github.com/eclipse-equinox/p2/assets/208716/041435a5-9c35-4e50-9207-bfceab3dca35)

This suggests that perhaps some type of retry is missing.

I know from past problems that the download server drops connections very quickly and this problem looks similar since we only fail because of what appear to be expired connections.

@laeubi

Any ideas of what I could try to change?

**Comment by laeubi** [comment]
I think in general there is no guarantee of what kind of exceptions are thrown in case of errors especially as here no byte are received at all. Maybe one can look for `EOFException` in the exception hierarchy...

If one want to get more fine grained exceptions the way would be to report it to the OpenJDK so then can throw a more specific exception.

**Comment by merks** [comment]
I would hope that a connection expired exception is handled by reestablishing the connection.  If I recall correctly in the previous implementation it was quite common for the reused connection to expire because download.eclipse.org expires them quickly.  I.e., I think it’s normal for a connection to expire when it’s reused. 

Consistently getting a handful of logged problems with the new implementation versus never having seen that with the old implementation seems like a regression and like something we should address.  Perhaps you see that differently?

**Comment by laeubi** [comment]
> I would hope that a connection expired exception is handled by reestablishing the connection.

What do you mean by "connection expired"? For me it looks like that the server is forcefully closing the connection in an unusual way or Oomph is using the Transport in an unusual way, Tycho uses Http11Client internally for a long time now and I have not seen any complains from users (including platform builds), but this *might* depend on used Java Version (for Tycho 17 or higher)

> I know from past problems that the download server drops connections very quickly and this problem looks similar since we only fail because of what appear to be expired connections.

Dropping the connection is really bad behavior, a server is expected to gracefully tear down connection using code 408:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/408

Beside that the client does not offer much compared to the bare minimum on transport level:
https://docs.oracle.com/en/java/javase/17/docs/api/java.net.http/java/net/http/HttpClient.Builder.html

so if you think some retry should be implemented it must be done on the ECF side, so e.g. a GET *should* be retry-able but for me it don't looks like an expired connection but more of a drop maybe because of too many requests from the same host or something like that.

**Comment by merks** [comment]
As I showed in https://github.com/eclipse-equinox/p2/issues/301#issuecomment-1630992750, I mean that the jdk.internal.net.http.common.ConnectionExpiredException is being thrown.  That's what's leading to this failure.  This is happening during a HEAD request not during a GET request.  It's super hard to debug with so many threads involved and it being timing related. 

It seems quite a few folks run into this problem, but workarounds such as using HTTP version 1.1 do not actually work:

https://www.google.com/search?q=HTTP%2F1.1+header+parser+received+no+byte

There are even somewhat-related JDK bugs such as this:

https://bugs.openjdk.org/browse/JDK-8299267

So I don't think this is a bug this new ECF implementation but rather poor handling within the JDK's httpclient implementation and possibly poor behavior or poor header information about connection expiration (keep-alive) from the download.eclipse.org server.

In any case, I don't see how to resolve this problem with any ease and you don't appear to be all that interested in the problem in the first place.   Furthermore, this is "just" logging from stats handling, MirrorRequest.collectStats(MirrorRequest.java:279), so failures are not actual very relevant (except maybe to the folks defining an update site) and retries might cause over counting.

**Comment by merks** [comment]
@laeubi 

Comparing behaviors of the two implementations again, I think this is a more sensible/appropriate fix because there are places that null the httpRequest that also in the other implmentation nulled the getMethod which was used as an isConnected guard in the other implementation:

![image](https://github.com/eclipse-equinox/p2/assets/208716/7f13ab27-30bd-4d54-b9d2-c814ab8deb34)

This needs to be fixed in ECF and we need a new ECF build with that fix.

Is there an ECF issue open against which to create a PR?

**Comment by laeubi** [comment]
> Is there an ECF issue open against which to create a PR?

I don't think you but you can just open a PR and reference the P2 issue that how I did it last time.

> As I showed in https://github.com/eclipse-equinox/p2/issues/301#issuecomment-1630992750, I mean that the jdk.internal.net.http.common.ConnectionExpiredException is being thrown.

I must have missed that no idea why JDT is not showing the exception :-\

> but workarounds such as using HTTP version 1.1 do not actually work

Yeah that would be undesirable, I hope we can use Http/2 in some cases and getting better performance from that.

> In any case, I don't see how to resolve this problem with any ease and you don't appear to be all that interested in the problem in the first place.

If we can handle this in the ECF code it would be good `HEAD` and `GET` both should be idempotent and could be retried.

**Comment by laeubi** [comment]
The jdk bug is meant to be fixed in Java 17: https://bugs.openjdk.org/browse/JDK-8299018 what JDK do you use for testing?

**Comment by merks** [comment]
That exception can't be found with Ctrl-Shift-T.  It's really very internal!  😜 

My vague impression is that the performance is very good.  Amazing how fast 300 files can be downloaded for an install!! It would be annoying to force 1.1, and it's useless anyway, so a moot point.

> The jdk bug is meant to be fixed in Java 17: https://bugs.openjdk.org/browse/JDK-8299018 what JDK do you use for testing?

I did update to 17.0.7 (latest available Adoptium release) and retest with that, though I still had the same issue.

Shall I proceed with a PR for the simple fix proposed in https://github.com/eclipse-equinox/p2/issues/301#issuecomment-1632289161 ?
